doc: add domain post mortem

trevnorris opened the pull request:
##### Checklist

- [x] documentation is changed or added
- [x] the commit message follows commit guidelines

##### Affected core subsystem(s)

doc

##### Description of change

The domain module did not live a long and fruitful life. Instead it was
ultimately deprecated and abandon. The domain postmortem addresses why
this happened so that it doesn't happen again in the future.

Also included is an example script that accompanies the postmortem
analysis.

R=@Fishrock123 
R=@nodejs/documentation

cjihrig commented on doc/topics/domain-postmortem.md:
```diff
@@ -0,0 +1,138 @@
+built-in http benchmark, `http_simple.js`, without domains it can handle over
+22,000 requests/second. Whereas if it's run with `NODE_USE_DOMAINS=1` that
+number drops down to under 17,000 requests/second. In this case there is only
+as single global domain. If we edit the benchmark so the http request callback
```
`as` -> `a`

eljefedelrodeodeljefe replied:
`a single domain`?

trevnorris replied:
global is more technically correct. though even more so if it said a single active global domain. since the domain is activated at startup and left active for the remainder of the script.

eljefedelrodeodeljefe commented on doc/topics/domain-postmortem.md:
```diff
@@ -0,0 +1,138 @@
+# domain Module Postmortem
```
nit `Domain`

evanlucas replied:
either that or wrap it in backticks `domain`?

eljefedelrodeodeljefe commented:
docs-wise LGTM with nits. Good write-up.

cjihrig commented on doc/topics/domain-postmortem.md:
```diff
@@ -0,0 +1,138 @@
+intercept the exceptions of unrelated code in a different module. Preventing
+the originator of the code from knowing about its own exceptions.
+
+Another is the how domains route errors automatically if no `'error'` handler
```
Starting with "Another" sounds kind of funny here. Maybe change it to "Domains also route errors..."

bengl replied:
Perhaps "Another issue is that domains route errors automatically...."

cjihrig commented on doc/topics/domain-postmortem.md:
```diff
@@ -0,0 +1,138 @@
+d1.run(() => setTimeout(() => {
+  const d2 = domain.create();
+  d2.bar = 43;
+  d2.on('error', (er) => print(er.message, domain._stack));
```
What is `print()`?

trevnorris replied:
heh. i have a vim macro that auto-adds `const print = process._rawDebug` to match d8. i'll fix that.

mscdex commented:
While on the topic, are there any thoughts about replacements and/or the way forward? Would it be beneficial to include at least a little information on that at the end of this document or no?

trevnorris commented:
@mscdex I'm hopeful that AsyncWrap can assume the functionality of continuous local storage. The postmortem intended to demonstrate the difficulty in properly implementing asynchronous error handling, and how anything mildly complex leads to a situation where reliable resource cleanup is impossible.

Since my initial inception of AsyncListener, and its error handling implementation, I've reiterated on how to make this more feasible. After the initial features are implemented I'll write up a design doc for review.

trevnorris commented:
All nits addressed.

@mscdex Would you like a small section about how we plan on moving forward now that this functionality has been deprecated? Would also like input from others on how this should be addressed.

benjamingr commented on doc/topics/domain-postmortem.md:
```diff
@@ -0,0 +1,138 @@
+`domain.enter()`. Which then acts as a catch-all for any exception in the
+future that couldn't be observed by the thrower. Allowing a module author to
+intercept the exceptions of unrelated code in a different module. Preventing
+the originator of the code from knowing about its own exceptions.
```
A code sample here would do _wonders_ to explain this to people considering domains.

trevnorris replied:
added.

benjamingr commented on doc/topics/domain-postmortem.md:
```diff
@@ -0,0 +1,138 @@
+seem useful at first, but once asynchronous calls are two or more modules deep
+and one of them doesn't include an error handler the creator of the domain will
+suddenly be catching unexpected exceptions, and the thrower's exception will go
+unnoticed by the author.
```
A code sample here would also do _wonders_ to explain this. Aside from that both examples are very good.

trevnorris replied:
added.

benjamingr commented on doc/topics/domain-postmortem.md:
```diff
@@ -0,0 +1,138 @@
+      setTimeout(() => {
+        throw new Error('AGAIN?');
+      });
+      throw new Error('ah crap')
```
Maybe not 'ah crap' and 'AGAIN?' but "inner" and "outer" ?

trevnorris replied:
I was trying so hard to slip "crap" into a doc w/o anyone noticing. :P 

"inner"/"outer" are more descriptive. Will use those.

benjamingr commented on doc/topics/domain-postmortem.md:
```diff
@@ -0,0 +1,138 @@
+domain API. All that domains offer is an exception aggregation mechanism. Even
+the potentially useful ability to propagate data with the domain is easily
+countered, in this example, by passing the needed resources as a function
+argument.
```
For what it's worth, most complaints I've seen about deprecating domains are about context and not error handling.

trevnorris replied:
are you referring to users using domains as a type of continuation local storage?

benjamingr replied:
Yes, very often people ask in StackOverflow for storage for the duration of a request. They complain about having to pass parameters to 6-7 functions.

In my opinion passing the parameters around a-la-dependency-injection is a much better and safer idea than 'request/continuation/"thread" local storage'  - but I'm just pointing out that people do that. That's just StackOverflow though, maybe it's a lot less common in the wild.

trevnorris replied:
Thanks for clarifying. I know it's being used, but only by users. Haven't come across a module that uses it yet (though I also haven't looked very hard). I'll be happy to make a note of it. Think I made a note of it, but was mostly just cumbersome than anything. e.g. if another module did the same within your async branch then the domains would stack and `process.domain` might not have the information you were looking for.

trevnorris replied:
section added for this topic.

benjamingr commented on doc/topics/domain-postmortem.md:
```diff
@@ -0,0 +1,138 @@
+any of its dependencies. Breaking down the script into its basic operations:
+
+- On new connection:
+  - Open a file on the file system
```
maybe `When a new connection happens, concurrently:` to illustrate that fact - concurrent resource cleanup is much harder than sequential cleanup and it might help illustrate the complexity.

benjamingr commented:
I left some comments @trevnorris - otherwise excellently written :clap:

mscdex commented:
@trevnorris It's up to you, but after reading all of it it just seemed to kind of leave me hanging at the end.

jasnell commented:
LGTM

trevnorris commented:
@benjamingr Added two examples, changed error message and added a new section, with example, on using domain for data propagation.

@mscdex See your point. I'll try to get something written up, but no promises.

benjamingr commented:
Very nice.

Fishrock123 commented:
I think `domain-resource-cleanup.js` should be `domain-resource-cleanup-example.js`

Or `examples/domain-resource-cleanup.js`, though I think `domain-resource-cleanup-example.js` should be fine.

Fishrock123 commented:
LGTM otherwise.

rvagg commented on doc/topics/domain-postmortem.md:
```diff
@@ -0,0 +1,289 @@
+
+### API Gaps
+
+While APIs based using `EventEmitter` can use `bind()` and errback style
```
"based using" -> "based on using"?

trevnorris replied:
fixed.

rvagg commented:
Nice work @trevnorris, so good to have all of this in one place that we can point to when it comes up.

I think this needs an intro to explain _why_ this doc exists in the first place, something along the lines of "The Domains API has been deprecated, this document explains some of the major problems observed with Domains and why the decision was made to scrap, rather than attempt to fix the API".

Another possible addition here could be a section on complexity and the overhead that Domains has introduced in maintenance of so many parts of core.

/cc @misterdjules who may have some good insight to add on this topic.

bengl commented on doc/topics/domain-postmortem.md:
```diff
@@ -0,0 +1,289 @@
+Even manually removing the connection via `d.remove(c)` does not prevent the
+connection's error from being automatically intercepted.
+
+A failure that plagues both error routing and exception handling are the
```
Grammar nit: "A failure ... are ..." <-- singular subject, plural verb.

trevnorris replied:
fixed

Fishrock123 commented:
> Another possible addition here could be a section on complexity and the overhead that Domains has introduced in maintenance of so many parts of core.

domains + timers = 😞

misterdjules commented:
@rvagg Thanks for the heads up! @trevnorris I think this is a great document. I found good descriptions for every issue I've run into in the past with domains.

The only minor observations and comments that I have are:

1) This document mentions that domains have been and still are difficult to maintain in core. I'm not sure this is inherent to domains, it may be a sign that the implementation and its tests were and are still not good enough. When working on fixing domains related issues, I often found that the tests suite covered only simple used cases, which made introducing regressions much easier, and understanding the intentions of domains much harder. I've always tried to add more tests when fixing bugs, and today I'm comfortable making changes to domains.

2) The description of the PR states that domains are "abandoned". I don't think this is accurate, because we still fix bugs in the domains module and we help users troubleshoot problems they have when using them.

3) It also states that "the domain module did not live a long and fruitful life.". I think this might be a bias from the perspective of someone who had to work in the implementation of domains, but it seems that a significant number of users have used domains for a long time, and some of them have been happy with it.

In other words I would encourage us to not underestimate the value of domains for some node users, even if in hindsight it appears to be a less than ideal solution for their problem.

trevnorris commented:
@misterdjules 

1) Difficult to maintain as in difficult to work around. Remember the `TryCatch` issue in `MakeCallback`? Domains added another several days to my debugging since they changed the behavior of how those cases would work. They make implementing new features, and at time fixing existing ones, more difficult simply because they exist.

2) Abandon was a poor choice of word. It does not exist in the document itself. Will change the commit message body.

3) This wording is based on the fact that domain was introduced in April 2012, and then deprecated in Dec 2014. The `dispose()` API didn't even make it a full year (deprecated in Mar 2013). The users that have been using domain since then are using a deprecated module.

trevnorris commented:
@Fishrock123 Renamed the file.

@mscdex I'm going to refrain from adding information about how we move forward ATM because of the flux of current proposals (e.g. zones and asyncwrap). I'll come back and add information after things have settled down a little.

Everyone, anything else?

mscdex commented:
@trevnorris Fair enough, but I wasn't even really looking for specifics so much as I was just some tidbit at the end about domains being (or at least the aim is to be) replaced by something else and that domains are not just something that was an experiment that is being removed and forgotten. It's not a big deal though.

benjamingr commented:
Still LGTM

rvagg commented:
I still wouldn't mind some intro text instead of launching straight into the technical issues, just an answer to: why does this document exist?

trevnorris commented:
@rvagg because I was asked to do it?  i'm honestly not sure what to put other than "domains are deprecated. this document explains why."

trevnorris commented:
@mscdex I've added a "Looking Ahead" section at the end. Hopefully that's along the lines of what you were looking for.

@rvagg I would be happy to add a prologue or such if you have more details of what you're looking for.

mscdex commented:
@trevnorris LGTM

trevnorris commented:
Thanks all. Landed in 4a74fc9776d.

TheAlphaNerd commented:
@trevnorris I've gone ahead and marked this don't land on v4

Please let me know if you think it should be backported to lts docs

trevnorris commented:
@TheAlphaNerd Sounds fine. This is only informative.
